### ISSUE TITLE
tie doesn't autoviv an aliased hash element

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7128]
Migrated from [rt.perl.org#27010](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=27010) (status was 'resolved')

Searchable as RT27010$

**Comment by p5pRT** [comment]
### From andrew@pimlott.net

#### Created by andrew@pimlott.net



The following amusing function works&#8203;:

&nbsp;    sub assign \{ $\_\[0\] = $\_\[1\] \}
&nbsp;    assign $foo => 1;
&nbsp;    assign $foo\{bar\} => 1;

In particular\, in the last line\, the hash element is autovivified by the
assignment to the alias $\_\[0\]\.  The following doesn't&#8203;:

&nbsp;    sub maitai \{ tie $\_\[0\] => 'MaiTai' \}
&nbsp;    maitai $foo;        \# ok
&nbsp;    maitai $foo\{bar\};   \# not ok

$foo gets tied as expected\, but $foo\{bar\} doesn't\.  Evidently\, whatever got
tied was dissociated from the hash element\.  This can be repaired with

&nbsp;    sub maitai \{ tie $\_\[0\]\, 'MaiTai' \}
&nbsp;    $foo\{bar\} = undef;
&nbsp;    maitai $foo\{bar\};   \# ok

However\, this doesn't do it&#8203;:

&nbsp;    sub maitai \{ $\_\[0\] = 'DUMMY'; tie $\_\[0\]\, 'MaiTai' \}
&nbsp;    maitai $foo\{bar\};   \# $foo\{bar\} eq 'DUMMY' \!\!

Last\,

&nbsp;    tie $foo\{bar\} => 'MaiTai';

works fine\.  So the problem appears to be autovivifying\, through an alias\,
in tie\.  It would be cool if this worked\.

Here is a tie implementation to test with&#8203;:

&nbsp;    package MaiTai;
&nbsp;    sub TIESCALAR \{ bless \{\} => MaiTai; \}
&nbsp;    sub FETCH \{ 'TIED' \}

Andrew

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.2:

Configured by Debian Project at Sat Nov 15 18:33:34 EST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 2) configuration:
  Platform:
    osname=linux, osvers=2.4.22-xfs+ti1211, archname=i386-linux-thread-multi
    uname='linux kosh 2.4.22-xfs+ti1211 #1 sat oct 25 10:11:37 est 2003 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8.2 -Darchlib=/usr/lib/perl/5.8.2 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.2 -Dsitearch=/usr/local/lib/perl/5.8.2 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.2 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3.2 (Debian)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so.5.8.2
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.2:
    /home/pimlott/lib/perl
    /home/pimlott/local/lib/perl
    /etc/perl
    /usr/local/lib/perl/5.8.2
    /usr/local/share/perl/5.8.2
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8.2
    /usr/share/perl/5.8.2
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.2:
    HOME=/home/pimlott
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/local/atria/bin:/usr/sbin:/sbin:/home/pimlott/bin:/usr/local/atria/bin:/usr/sbin:/sbin:/home/pimlott/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games
    PERL5LIB=/home/pimlott/lib/perl:/home/pimlott/local/lib/perl
    PERL_BADLANG (unset)
    SHELL=/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @ysth

On Mon\, Feb 23\, 2004 at 07&#8203;:18&#8203;:23PM \-0000\, "andrew@&#8203;pimlott\.net \(via RT\)" wrote&#8203;:
>     sub maitai \{ tie $\_\[0\] => 'MaiTai' \}
>     maitai $foo;        \# ok
>     maitai $foo\{bar\};   \# not ok
> 
> $foo gets tied as expected\, but $foo\{bar\} doesn't\.  Evidently\, whatever got
> tied was dissociated from the hash element\.  This can be repaired with
> 
>     sub maitai \{ tie $\_\[0\]\, 'MaiTai' \}
>     $foo\{bar\} = undef;
>     maitai $foo\{bar\};   \# ok

My thoughts on this&#8203;:

pp\_tie \(or sv\_magic\*?\) could handle this for nonexistent elements
\(LvTYPE=='y'\) \(which change should also go into maint\)\.  A side effect
of the tie would be to create the element\.

For other types of passable lvalues \(vec\, pos\, scalar keys\, 3\-arg
substr\, element of magic array or hash\) it should warn and the tie
fail\.  In no case should the tie actually add magic to a PVLV\.
Presumably having the tie fail should go in maint also\, but not
necessarily the new warning\.  Nicholas?

Any other opinions?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Feb 23 11&#8203;:18&#8203;:23 2004\, andrew@&#8203;pimlott\.net wrote&#8203;:
> 
> This is a bug report for perl from andrew@&#8203;pimlott\.net\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> The following amusing function works&#8203;:
> 
>     sub assign \{ $\_\[0\] = $\_\[1\] \}
>     assign $foo => 1;
>     assign $foo\{bar\} => 1;
> 
> In particular\, in the last line\, the hash element is autovivified by
>    the
> assignment to the alias $\_\[0\]\.  The following doesn't&#8203;:
> 
>     sub maitai \{ tie $\_\[0\] => 'MaiTai' \}
>     maitai $foo;        \# ok
>     maitai $foo\{bar\};   \# not ok
> 
> $foo gets tied as expected\, but $foo\{bar\} doesn't\.  Evidently\,
>    whatever got
> tied was dissociated from the hash element\.  This can be repaired with
> 
>     sub maitai \{ tie $\_\[0\]\, 'MaiTai' \}
>     $foo\{bar\} = undef;
>     maitai $foo\{bar\};   \# ok
> 
> However\, this doesn't do it&#8203;:
> 
>     sub maitai \{ $\_\[0\] = 'DUMMY'; tie $\_\[0\]\, 'MaiTai' \}
>     maitai $foo\{bar\};   \# $foo\{bar\} eq 'DUMMY' \!\!
> 
> Last\,
> 
>     tie $foo\{bar\} => 'MaiTai';
> 
> works fine\.  So the problem appears to be autovivifying\, through an
>    alias\,
> in tie\.  It would be cool if this worked\.
> 
> Here is a tie implementation to test with&#8203;:
> 
>     package MaiTai;
>     sub TIESCALAR \{ bless \{\} => MaiTai; \}
>     sub FETCH \{ 'TIED' \}

I am in the middle of fixing this\, but I have run into something
interesting\.

Obviously sub \{ tie $\_\[0\]\.\.\. \} \->\($h\{nonexistent\}\) should vivify the
element\.

But should tied $\_\[0\] vivify it?  What about untie?

These two functions do already autovivify in another way&#8203;:

$ \./perl \-lIlib \-e 'tied %$\_; print $\_'
HASH\(0x7fa2640052b8\)
$ \./perl \-lIlib \-e 'untie %$\_; print $\_'
HASH\(0x7f8d688052b8\)

Should they?

It makes some sense for untie to put its argument in vivifying context\,
since it modifies it\, though it would work just as well without
autovivifying\.

To me\, it doesn’t make sense that tied\(\) should autovivify at all\.

However tied %$\_ works\, should the same rules apply to tied $\_\[0\] where
$\_\[0\] represents a nonexistent element?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Jul 15 23&#8203;:28&#8203;:26 2013\, sprout wrote&#8203;:
> However tied %$\_ works\, should the same rules apply to tied $\_\[0\] where
> $\_\[0\] represents a nonexistent element?

In other words\, am I perhaps conflating two unrelated issues here?

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Mon Jul 15 23&#8203;:28&#8203;:26 2013\, sprout wrote&#8203;:
> On Mon Feb 23 11&#8203;:18&#8203;:23 2004\, andrew@&#8203;pimlott\.net wrote&#8203;:
> > 
> > This is a bug report for perl from andrew@&#8203;pimlott\.net\,
> > generated with the help of perlbug 1\.34 running under perl v5\.8\.2\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > The following amusing function works&#8203;:
> > 
> >     sub assign \{ $\_\[0\] = $\_\[1\] \}
> >     assign $foo => 1;
> >     assign $foo\{bar\} => 1;
> > 
> > In particular\, in the last line\, the hash element is autovivified by
> >    the
> > assignment to the alias $\_\[0\]\.  The following doesn't&#8203;:
> > 
> >     sub maitai \{ tie $\_\[0\] => 'MaiTai' \}
> >     maitai $foo;        \# ok
> >     maitai $foo\{bar\};   \# not ok
> > 
> > $foo gets tied as expected\, but $foo\{bar\} doesn't\.  Evidently\,
> >    whatever got
> > tied was dissociated from the hash element\.  This can be repaired with
> > 
> >     sub maitai \{ tie $\_\[0\]\, 'MaiTai' \}
> >     $foo\{bar\} = undef;
> >     maitai $foo\{bar\};   \# ok
> > 
> > However\, this doesn't do it&#8203;:
> > 
> >     sub maitai \{ $\_\[0\] = 'DUMMY'; tie $\_\[0\]\, 'MaiTai' \}
> >     maitai $foo\{bar\};   \# $foo\{bar\} eq 'DUMMY' \!\!
> > 
> > Last\,
> > 
> >     tie $foo\{bar\} => 'MaiTai';
> > 
> > works fine\.  So the problem appears to be autovivifying\, through an
> >    alias\,
> > in tie\.  It would be cool if this worked\.
> > 
> > Here is a tie implementation to test with&#8203;:
> > 
> >     package MaiTai;
> >     sub TIESCALAR \{ bless \{\} => MaiTai; \}
> >     sub FETCH \{ 'TIED' \}
> 
> I am in the middle of fixing this\, but I have run into something
> interesting\.
> 
> Obviously sub \{ tie $\_\[0\]\.\.\. \} \->\($h\{nonexistent\}\) should vivify the
> element\.
> 
> But should tied $\_\[0\] vivify it?  What about untie?
> 
> These two functions do already autovivify in another way&#8203;:
> 
> $ \./perl \-lIlib \-e 'tied %$\_; print $\_'
> HASH\(0x7fa2640052b8\)
> $ \./perl \-lIlib \-e 'untie %$\_; print $\_'
> HASH\(0x7f8d688052b8\)
> 
> Should they?
> 
> It makes some sense for untie to put its argument in vivifying context\,
> since it modifies it\, though it would work just as well without
> autovivifying\.
> 
> To me\, it doesn’t make sense that tied\(\) should autovivify at all\.
> 
> However tied %$\_ works\, should the same rules apply to tied $\_\[0\] where
> $\_\[0\] represents a nonexistent element?

I have fixed this bug in commit 13733cde\.  I chose not to vivify
elements in tied and untie\.  Since there is no precedent for this sort
of thing\, I picked one behaviour and went with it\.  I think preferring
not to vivify when it is not strictly necessary is better\.  I did not
touch the existing vivification of references\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From andrew@pimlott.net

Excerpts from Father Chrysostomos via RT's message of Tue Jul 16 00&#8203;:29&#8203;:12 \-0700 2013&#8203;:
> I have fixed this bug in commit 13733cde\.

Thanks for making the magic alive\. :\-\)

Andrew
